tests/integration/cache_test.go

```diff
@@ -160,98 +199,160 @@ func testWatch(t *testing.T, kv clientv3.KV, watcher Watcher) {
 		},
 		{
 			name:       "Watch from empty key",
 			key:        "",
-			opts:       []clientv3.OpOption{clientv3.WithFromKey()},
-			wantEvents: []*clientv3.Event{event1PutFooA, event2PutFooB, event3DeleteFooA, event4PutFooA, event5DeleteFooB, event6PutFooC, event7PutFooBar, event8PutFooBaz, event9PutFooYoo, event10PutZoo},
+			opts:       []clientv3.OpOption{clientv3.WithFromKey(), clientv3.WithRev(2)},
+			wantEvents: []*clientv3.Event{rev2PutFooA, rev3PutFooB, rev4DeleteFooA, rev5PutFooA, rev5DeleteFooB, rev6PutFooC, rev7PutFooBar, rev8PutFooBaz, rev9PutFooYoo, rev10PutZoo, rev11PutFooFuture, rev12PutFooTx1, rev12DeleteFooFuture, rev12PutFooTx2},
 		},
 		{
 			name:       "Watch from non-existent key /doesnotexist",
 			key:        "/doesnotexist",
-			opts:       []clientv3.OpOption{clientv3.WithFromKey()},
-			wantEvents: []*clientv3.Event{event1PutFooA, event2PutFooB, event3DeleteFooA, event4PutFooA, event5DeleteFooB, event6PutFooC, event7PutFooBar, event8PutFooBaz, event9PutFooYoo, event10PutZoo},
+			opts:       []clientv3.OpOption{clientv3.WithFromKey(), clientv3.WithRev(2)},
+			wantEvents: []*clientv3.Event{rev2PutFooA, rev3PutFooB, rev4DeleteFooA, rev5PutFooA, rev5DeleteFooB, rev6PutFooC, rev7PutFooBar, rev8PutFooBaz, rev9PutFooYoo, rev10PutZoo, rev11PutFooFuture, rev12PutFooTx1, rev12DeleteFooFuture, rev12PutFooTx2},
+		},
+		{
+			name:       "Watch from rev 4 with single key /foo/a",
+			key:        "/foo/a",
+			opts:       []clientv3.OpOption{clientv3.WithRev(4)},
+			wantEvents: []*clientv3.Event{rev4DeleteFooA, rev5PutFooA},
+		},
+		{
+			name:       "Watch from rev 6 with single key /foo/a",
+			key:        "/foo/a",
+			opts:       []clientv3.OpOption{clientv3.WithRev(6)},
+			wantEvents: nil,
+		},
+		{
+			name: "Watch from rev 5 with prefix /foo",
+			key:  "/foo",
+			opts: []clientv3.OpOption{clientv3.WithPrefix(), clientv3.WithRev(5)},
+			wantEvents: []*clientv3.Event{
+				rev5PutFooA, rev5DeleteFooB, rev6PutFooC, rev7PutFooBar, rev8PutFooBaz, rev9PutFooYoo, rev11PutFooFuture, rev12PutFooTx1, rev12DeleteFooFuture, rev12PutFooTx2,
+			},
+		},
+		{
+			name: "Watch from rev 10 with prefix /foo",
+			key:  "/foo",
+			opts: []clientv3.OpOption{clientv3.WithPrefix(), clientv3.WithRev(10)},
+			wantEvents: []*clientv3.Event{
+				rev11PutFooFuture, rev12PutFooTx1, rev12DeleteFooFuture, rev12PutFooTx2,
+			},
+		},
+		{
+			name: "Watch from rev 4 with range [/foo/a, /foo/c)",
+			key:  "/foo/a",
+			opts: []clientv3.OpOption{clientv3.WithRange("/foo/c"), clientv3.WithRev(4)},
+			wantEvents: []*clientv3.Event{
+				rev4DeleteFooA, rev5PutFooA, rev5DeleteFooB, rev7PutFooBar, rev8PutFooBaz,
+			},
+		},
+		{
+			name:       "Watch from rev 11 with single key /foo/future",
+			key:        "/foo/future",
+			opts:       []clientv3.OpOption{clientv3.WithRev(11)},
```

Comment:
```suggestion
			name:       "Watch with single key /foo/future",
			key:        "/foo/future",
			opts:       nil,
```
For watcher on revisions written after we open a watch please drop WithRev